I'll start by examining the data in the file to understand the available features and structure. After that, I'll help you build a framework for evaluating and comparing regression models based on your outlined objective. Let me load and review the contents of the dataset first.

In [ ]:
import pandas as pd

# Load the dataset to examine its structure and contents
file_path = '/content/Student_Performance.csv'
data = pd.read_csv(file_path)

# Display the first few rows and a summary of the dataset to understand its structure
data.head(), data.describe()


(   Hours Studied  Previous Scores Extracurricular Activities  Sleep Hours  \
 0              7               99                        Yes            9   
 1              4               82                         No            4   
 2              8               51                        Yes            7   
 3              5               52                        Yes            5   
 4              7               75                         No            8   
 
    Sample Question Papers Practiced  Performance Index  
 0                                 1               91.0  
 1                                 2               65.0  
 2                                 2               45.0  
 3                                 2               36.0  
 4                                 5               66.0  ,
        Hours Studied  Previous Scores   Sleep Hours  \
 count   10000.000000     10000.000000  10000.000000   
 mean        4.992900        69.445700      6.530600   
 std         

The dataset includes the following features:

1.Hours Studied: Continuous numerical feature.

2.Previous Scores: Continuous numerical feature representing past academic scores.

3.Extracurricular Activities: Categorical feature ("Yes" or "No").

4.Sleep Hours: Continuous numerical feature indicating hours of sleep.

5.Sample Question Papers Practiced: Continuous numerical feature.

The target variable is Performance Index, a continuous value indicating student performance.

To proceed, I’ll:

1.Preprocess the data, including encoding categorical features.

2.Split the data into training and test sets.

3.Develop and evaluate different regression models (e.g., Linear Regression, Random Forest Regressor).

4.Compare models based on accuracy and other metrics to select the best-performing model.

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

# Encode the categorical variable 'Extracurricular Activities'
data['Extracurricular Activities'] = LabelEncoder().fit_transform(data['Extracurricular Activities'])

# Define features and target variable
X = data.drop(columns=['Performance Index'])
y = data['Performance Index']

# Split the dataset into training and test sets (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Display the shapes of the train and test sets to confirm the split
X_train.shape, X_test.shape, y_train.shape, y_test.shape


((8000, 5), (2000, 5), (8000,), (2000,))

The dataset has been split into training and test sets, with 8,000 samples for training and 2,000 for testing. Now, I'll proceed to develop and evaluate several regression models, starting with common choices like Linear Regression and Random Forest Regressor. This will give us an initial comparison based on metrics such as Mean Absolute Error (MAE) and R-squared (R²).

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, r2_score

# Initialize models
linear_reg = LinearRegression()
random_forest = RandomForestRegressor(random_state=42)

# Train models
linear_reg.fit(X_train, y_train)
random_forest.fit(X_train, y_train)

# Make predictions on the test set
y_pred_linear = linear_reg.predict(X_test)
y_pred_rf = random_forest.predict(X_test)

# Evaluate models
linear_metrics = {
    "MAE": mean_absolute_error(y_test, y_pred_linear),
    "R2": r2_score(y_test, y_pred_linear)
}

rf_metrics = {
    "MAE": mean_absolute_error(y_test, y_pred_rf),
    "R2": r2_score(y_test, y_pred_rf)
}

linear_metrics, rf_metrics


({'MAE': 1.6111213463123044, 'R2': 0.9889832909573145},
 {'MAE': 1.8147207361111108, 'R2': 0.9860792088401176})

The initial model evaluations show the following performance:

1.Linear Regression:

Mean Absolute Error (MAE): 1.61

R-squared (R²): 0.989

2.Random Forest Regressor:

Mean Absolute Error (MAE): 1.81

R-squared (R²): 0.986

Both models perform well, with Linear Regression slightly outperforming Random Forest on MAE and R². Based on these results, the Linear Regression model may be a strong candidate for the final model due to its simplicity and competitive performance.

In [ ]:
def predict_performance_index(hours_studied, previous_scores, extracurricular_activities, sleep_hours, sample_papers_practiced):
    """
    Predicts the Performance Index of a student based on the given features.

    Parameters:
    - hours_studied (float): Number of hours the student studied.
    - previous_scores (float): Student's previous scores.
    - extracurricular_activities (str): "Yes" if the student participates in extracurricular activities, otherwise "No".
    - sleep_hours (float): Number of hours the student sleeps.
    - sample_papers_practiced (float): Number of sample question papers practiced.

    Returns:
    - Predicted Performance Index (float).
    """
    # Encode the extracurricular activities feature
    extracurricular_encoded = 1 if extracurricular_activities.lower() == "yes" else 0

    # Arrange input data for prediction
    input_data = [[hours_studied, previous_scores, extracurricular_encoded, sleep_hours, sample_papers_practiced]]

    # Predict and return the performance index
    predicted_index = linear_reg.predict(input_data)
    return predicted_index[0]

# Example usage: predict the performance index for a new data point
example_prediction = predict_performance_index(6, 85, "Yes", 7, 4)
example_prediction


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


74.35148710879079